WalletWasabi/Extensions/NBitcoinExtensions.cs

```diff
@@ -374,5 +374,33 @@ void Walk(TransactionDependencyNode node)
 				nodes = parentCounter.Where(x => x.Value == 0).Select(x => x.Key).Distinct().ToArray();
 			}
 		}
+
+		public static ScriptPubKeyType? GetInputsScriptPubKeyType(this PSBT psbt)
+		{
+			if (!psbt.IsAllFinalized() || psbt.Inputs.Any(i => i.WitnessUtxo is null))
+			{
+				throw new InvalidOperationException("The psbt should be finalized with witness information");
+			}
+
+			var coinsPerTypes = psbt.Inputs.Select(i =>
+			{
+				return ((PSBTCoin)i, i.GetInputScriptPubKeyType());
+			}).GroupBy(o => o.Item2, o => o.Item1).ToArray();
+
+			return coinsPerTypes.Length != 1 ? (default) : coinsPerTypes[0].Key;
+		}
+
+		public static ScriptPubKeyType? GetInputScriptPubKeyType(this PSBTInput i)
+		{
+			if (i.WitnessUtxo.ScriptPubKey.IsScriptType(ScriptType.P2WPKH))
+			{
+				return ScriptPubKeyType.Segwit;
+			}
+
+			return i.WitnessUtxo.ScriptPubKey.IsScriptType(ScriptType.P2SH) &&
+				i.FinalScriptWitness.ToScript().IsScriptType(ScriptType.P2WPKH)
+				? (ScriptPubKeyType?)ScriptPubKeyType.SegwitP2SH
+				: null as ScriptPubKeyType?;
```

Comment:
There is this cast `(ScriptPubKeyType?)`, so I think it would be better to write it like this:
```suggestion
			if (i.WitnessUtxo.ScriptPubKey.IsScriptType(ScriptType.P2SH) && i.FinalScriptWitness.ToScript().IsScriptType(ScriptType.P2WPKH))
			{
				return ScriptPubKeyType.SegwitP2SH;
			}
			
			return null as ScriptPubKeyType?;
```

Reply:
Agree